rest/importtest/import_test.go

```diff
@@ -73,6 +73,42 @@ func TestImportFeed(t *testing.T) {
 	assert.Equal(t, int64(1), rt.GetDatabase().DbStats.SharedBucketImportStats.ImportCount.Value())
 }
 
+func TestImportFeedWithRecursiveSyncFunction(t *testing.T) {
+
+	base.SkipImportTestsIfNotEnabled(t)
+
+	rtConfig := rest.RestTesterConfig{
+		SyncFn:     `function access(doc) { access("foo", "bar"); }`,
+		AutoImport: base.Ptr(true),
+	}
+
+	rt := rest.NewRestTester(t, &rtConfig)
+	defer rt.Close()
+	dataStore := rt.GetSingleDataStore()
+
+	base.SetUpTestLogging(t, base.LevelDebug, base.KeyImport, base.KeyCRUD)
+
+	// Create doc via the SDK
+	mobileKey := t.Name()
+	mobileBody := make(map[string]interface{})
+	mobileBody["channels"] = "ABC"
+	_, err := dataStore.Add(mobileKey, 0, mobileBody)
+	assert.NoError(t, err, "Error writing SDK doc")
+
+	// Wait for import
+	err = rt.WaitForCondition(func() bool {
+		return rt.GetDatabase().DbStats.SharedBucketImportStats.ImportErrorCount.Value() == 1
+	})
+	require.NoError(t, err)
+
+	// Attempt to get the document via Sync Gateway.
+	response := rt.SendAdminRequest("GET", "/{{.keyspace}}/"+mobileKey, "")
+	assert.Equal(t, 404, response.Code)
```

Comment:
I know this code is copypasta from another test but I think writing it in a different way would be shorter and also fails on debug. `Add` will return no error if the document already exists, so to use `Add` robustly, you have to check the `added=true` return value. You can use `Set` to make sure the document exists. Using `base.RequireWaitForStat` is type safe and we can change to do better error reporting eventually.

```suggestion
	mobileKey := t.Name()
	require.NoError(t, dataStore.Set(mobileKey, 0, nil, db.Body{"channels": "ABC"}))

	// Wait for import
	base.RequireWaitForStat(t, rt.GetDatabase().DbStats.SharedBucketImportStats.ImportCount.Value, 1)

	// Attempt to get the document via Sync Gateway.
	response := rt.SendAdminRequest("GET", "/{{.keyspace}}/"+mobileKey, "")
	rest.RequireStatus(t, response, http.StatusOK)
```

Reply:
Will use wait for stat helpers, but I don't agree with using Set to hide the fact that doc may exist in bucket. Is it not a good assumption that a test bucket is empty? Why not?